node_exporter.go

```diff
@@ -27,51 +27,116 @@ import (
 	"gopkg.in/alecthomas/kingpin.v2"
 )
 
-func init() {
-	prometheus.MustRegister(version.NewCollector("node_exporter"))
+// handler wraps an unfiltered http.Handler but uses a filtered handler,
+// created on the fly, if filtering is requested. Create instances with
+// newHandler.
+type handler struct {
+	unfilteredHandler http.Handler
+	// exporterMetricsRegistry is a separate registry for the metrics about
+	// the exporter itself.
+	exporterMetricsRegistry *prometheus.Registry
+	includeExporterMetrics  bool
 }
 
-func handler(w http.ResponseWriter, r *http.Request) {
+func newHandler(includeExporterMetrics bool) *handler {
+	h := &handler{
+		exporterMetricsRegistry: prometheus.NewRegistry(),
+		includeExporterMetrics:  includeExporterMetrics,
+	}
+	if h.includeExporterMetrics {
+		h.exporterMetricsRegistry.MustRegister(
+			prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}),
+			prometheus.NewGoCollector(),
+		)
+	}
+	if innerHandler, err := h.makeInnerHandler(); err != nil {
+		log.Fatalf("Couldn't create metrics handler: %s", err)
+	} else {
+		h.unfilteredHandler = innerHandler
+	}
+	return h
+}
+
+// ServeHTTP implements http.Handler.
+func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	filters := r.URL.Query()["collect[]"]
 	log.Debugln("collect query:", filters)
 
-	nc, err := collector.NewNodeCollector(filters...)
+	if len(filters) == 0 {
+		// No filters, use the prepared unfiltered handler.
+		h.unfilteredHandler.ServeHTTP(w, r)
+		return
+	}
+	// To serve filtered metrics, we create a filtering handler on the fly.
+	filteredHandler, err := h.makeInnerHandler(filters...)
 	if err != nil {
-		log.Warnln("Couldn't create", err)
+		log.Warnln("Couldn't create filtered metrics handler:", err)
 		w.WriteHeader(http.StatusBadRequest)
-		w.Write([]byte(fmt.Sprintf("Couldn't create %s", err)))
+		w.Write([]byte(fmt.Sprintf("Couldn't create filtered metrics handler: %s", err)))
 		return
 	}
+	filteredHandler.ServeHTTP(w, r)
+}
 
-	registry := prometheus.NewRegistry()
-	err = registry.Register(nc)
+// makeInnerHandler is meant to be called exactly once with empty filters (in
+// which case it will log all the colletors enabled via command-line flags). It
+// should only be called again later if a filtering handler has to be created on
+// the fly.
+func (h *handler) makeInnerHandler(filters ...string) (http.Handler, error) {
```

Comment:
What about naming this `newFilteredHandler()`?

Reply:
Oh, I see, it's used more than once. I still think `new` rather than `make` is more sensible.

Reply:
I think `make` is better than `new` since the later suggest that it's a constructor for a type `FilteredHandler`. Could be called just `filteredHandler()` but either way, I think it's perfectly fine as is.

Reply:
Thanks for all the input. Naming is a bit problematic here indeed.

I didn't want to call it anything with “filtered” as that would read weird later when it is used without argument to get an unfiltered handler. So I used “inner” is a neutral term.

However, “make” vs “new” is interesting. I picked “make” here because “new” implies that a pointer is returned. However, an interface is inherently a pointer type, so I think `newInnerHandler` makes most sense here.

However, one could also just use `innerHandler`. It would be similar to the naming of `promhttp.HandlerFor`.
I think I'll go for the latter. It also seems to be in agreement (or at least not in contradiction) with all your comments.